packages/flutter_tools/test/base/build_test.dart

```diff
@@ -0,0 +1,192 @@
+final Matcher throwsBuildException = throwsA(const isInstanceOf<BuildException>());
+
+void main() {
+  group('BuildAction', () {
+    group('analyze', () {
+      test('throws BuildException if action declares the same file as input and output', () async {
+        await fs.file('a.dart').create();
+        await fs.file('b.dart').create();
+        final BuildAction action = new BuildAction(
+          inputPaths: <String>['a.dart', 'b.dart'].toSet(),
+          outputPaths: <String>['a.dart'].toSet(),
+          buildFunction: (Set<File> inputs) async {},
+        );
+        expect(() async => await action.build(), throwsBuildException);
+      });
+
+      test('does not throw BuildException if action declares no outputs that are also inputs', () async {
+        await fs.file('a.dart').create();
+        await fs.file('b.dart').create();
+        final BuildAction action = new BuildAction(
+          inputPaths: <String>['a.dart', 'b.dart'].toSet(),
+          outputPaths: <String>['c.dart'].toSet(),
+          buildFunction: (Set<File> inputs) async {
+            fs.file('c.dart').create();
+          },
+        );
+        action.build();
+      });
+    });
+
+    group('build', () {
+      MemoryFileSystem fs;
+
+      setUp(() {
+        fs = new MemoryFileSystem();
+      });
+
+      testUsingContext('throws if any input file does not exist', () async {
+        await fs.file('a.dart').create();
+        final BuildAction action = new BuildAction(
+          inputPaths: <String>['a.dart', 'b.dart'].toSet(),
+          outputPaths: <String>['c.dart'].toSet(),
+          buildFunction: (Set<File> files) async {
+            await fs.file('c.dart').create();
+          },
+        );
+        expect(() async => await action.build(), throwsBuildException);
+      }, overrides: <Type, Generator>{ FileSystem: () => fs });
+
+      testUsingContext('throws if build action does not emit output', () async {
+        await fs.file('a.dart').create();
+        await fs.file('b.dart').create();
+        final BuildAction action = new BuildAction(
+          inputPaths: <String>['a.dart', 'b.dart'].toSet(),
+          outputPaths: <String>['c.dart'].toSet(),
+          buildFunction: (Set<File> files) async {},
+        );
+        expect(() async => await action.build(), throwsBuildException);
+      }, overrides: <Type, Generator>{ FileSystem: () => fs });
+
+      testUsingContext('executes build action if all inputs exist', () async {
+        await fs.file('a.dart').create();
+        await fs.file('b.dart').create();
+        final BuildAction action = new BuildAction(
+          inputPaths: <String>['a.dart', 'b.dart'].toSet(),
+          outputPaths: <String>['c.dart'].toSet(),
+          buildFunction: (Set<File> files) async {
+            await fs.file('c.dart').create();
+          },
+        );
+        await action.build();
+        expect(await fs.file('c.dart').exists(), isTrue);
```

Comment:
`package:file/testing.dart` has an `exists` matcher (and some other handy matchers)

https://www.dartdocs.org/documentation/file/2.3.2/testing/testing-library.html